internal/generate/crd/crd.go

```diff
@@ -0,0 +1,342 @@
+
+// newCRDForResource constructs a barebones CRD using data in resource.
+func newCRDForResource(r scaffold.Resource) *apiextv1beta1.CustomResourceDefinition {
+	crd := &apiextv1beta1.CustomResourceDefinition{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: apiextv1beta1.SchemeGroupVersion.String(),
+			Kind:       "CustomResourceDefinition",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%s.%s", r.Resource, r.FullGroup),
+		},
+		Spec: apiextv1beta1.CustomResourceDefinitionSpec{
+			Group: r.FullGroup,
+			Scope: apiextv1beta1.NamespaceScoped,
+			Versions: []apiextv1beta1.CustomResourceDefinitionVersion{
+				{Name: r.Version, Served: true, Storage: true},
+			},
+			Subresources: &apiextv1beta1.CustomResourceSubresources{
+				Status: &apiextv1beta1.CustomResourceSubresourceStatus{},
+			},
+		},
+	}
+	setCRDNamesForResource(crd, r)
+	return crd
+}
+
+// setCRDNamesForResource creates a full set of CRD names using data in resource.
+func setCRDNamesForResource(crd *apiextv1beta1.CustomResourceDefinition, r scaffold.Resource) {
+	if crd.Spec.Names.Kind == "" {
+		crd.Spec.Names.Kind = r.Kind
+	}
+	if crd.Spec.Names.ListKind == "" {
+		crd.Spec.Names.ListKind = r.Kind + "List"
+	}
+	if crd.Spec.Names.Plural == "" {
+		crd.Spec.Names.Plural = r.Resource
+	}
+	if crd.Spec.Names.Singular == "" {
+		crd.Spec.Names.Singular = r.LowerKind
+	}
+}
+
+// setCRDStorageVersion sets exactly one version's storage field to true if
+// one is not already set.
+func setCRDStorageVersion(crd *apiextv1beta1.CustomResourceDefinition) {
+	if len(crd.Spec.Versions) == 0 {
+		return
+	}
+	for _, ver := range crd.Spec.Versions {
+		if ver.Storage {
+			return
+		}
+	}
+	// Set the first element in spec.versions to be the storage version.
+	log.Infof("Setting CRD %q storage version to %s", crd.GetName(), crd.Spec.Versions[0].Name)
+	crd.Spec.Versions[0].Storage = true
+}
+
+// checkCRDVersions ensures version(s) generated for a CRD are in valid format.
+// From the Kubernetes CRD docs:
+//
+// The version field is deprecated and optional, but if it is not empty,
+// it must match the first item in the versions field.
+func checkCRDVersions(crd *apiextv1beta1.CustomResourceDefinition) error {
+	singleVer := crd.Spec.Version != ""
+	multiVers := len(crd.Spec.Versions) > 0
+	if singleVer {
+		if !multiVers {
+			log.Warnf("CRD %s: spec.version is deprecated and should be migrated to spec.versions", crd.Spec.Names.Kind)
+		} else if crd.Spec.Version != crd.Spec.Versions[0].Name {
+			return errors.Errorf("spec.version %s must be the first element in spec.versions for CRD %s", crd.Spec.Version, crd.Spec.Names.Kind)
+		}
+	}
+
+	var hasStorageVer bool
+	for _, ver := range crd.Spec.Versions {
+		// There must be exactly one version flagged as a storage version.
+		if ver.Storage {
+			if hasStorageVer {
+				return errors.Errorf("spec.versions cannot have more than one storage version for CRD %s", crd.Spec.Names.Kind)
+			}
+			hasStorageVer = true
+		}
+	}
```

Comment:
```suggestion
    if ver.Storage && hasStorageVer {
	return errors.Errorf("spec.versions cannot have more than one storage version for CRD %s", crd.Spec.Names.Kind)
    }
    hasStorageVer = true
}
```

Reply:
Maybe ^. But is hard to understand this func logic.

Reply:
I'm trying not to change much of this logic since this is copied from existing, working code. Also if `ver.Storage` was false in your suggestion `hasStorageVer` would be set to true.